File: 프로그래머스/1/1845. 폰켓몬/폰켓몬.js
function solution(nums) {
    const table = nums.reduce((acc, num) => {
        if (!acc.has(num)) acc.set(num, 0);
        acc.set(num, acc.get(num) + 1);
        return acc;
    }, new Map());
    return [ ...table ].reduce((acc, pair) => {
        if (acc == nums.length / 2) {
            return acc;
        }
        return acc + 1;
    }, 0);
}

// 각 요소의 출현 횟수를 카운트
// key를 바탕으로 각 key별로 1 추가
// 2/n에 도달한 경우 더이상 추가 X
// 시간복잡도: O(n)

File: 프로그래머스/1/1845. 폰켓몬/폰켓몬.py
def solution(nums):
    pokemons = set()
    for num in nums:
        if num not in pokemons:
            pokemons.add(num)
    
    return min(len(nums) // 2, len(pokemons))

# 현재 선택한 포켓몬 종류 중에
#  없으면 담고
#  있으면 스킵
#  만약 nums/2만큼 담았으면 그만 담기